src/libstd/sys/unix/fd.rs

```diff
@@ -51,12 +51,21 @@ impl FileDesc {
         Ok(ret as usize)
     }
 
+    #[cfg(not(target_env = "newlib"))]
     pub fn set_cloexec(&self) {
         unsafe {
             let ret = c::ioctl(self.fd, c::FIOCLEX);
             debug_assert_eq!(ret, 0);
         }
     }
+    #[cfg(target_env = "newlib")]
+    pub fn set_cloexec(&self) {
+        unsafe {
+            let previous = c::fnctl(self.fd, c::F_GETFD);
+            let ret = c::fnctl(self.fd, c::F_SETFD, previous | c::FD_CLOEXEC);
```

Comment:
Newlib *can* have `ioctl` (for instance, linux has it), but not every platform provides one. In NaCl's case, `nacl_io` handles all `ioctl` and `fnctl` operations in place of `libc` (specifically, `nacl_io` provides a wrapper around the PPAPI IRT, allowing use of the usual set of file descriptor operations from within Chrome's sandbox), but it expects `FD_CLOEXEC` to be set via `fnctl` instead of `ioctl`. Also, `FIOCLEX` isn't defined for any platform.